onnxruntime/core/framework/kernel_registry_manager.cc

```diff
@@ -14,63 +14,105 @@ Status KernelRegistryManager::CreateKernel(const onnxruntime::Node& node,
                                            const IExecutionProvider& execution_provider,
                                            const SessionState& session_state,
                                            /*out*/ std::unique_ptr<OpKernel>& op_kernel) const {
-  std::lock_guard<OrtMutex> lock(lock_);
-  if (kernel_registries_.empty()) {
-    return Status(ONNXRUNTIME, FAIL, "Kernel not found.");
+  const std::string& ptype = node.GetExecutionProviderType();
+  if (ptype.empty()) {
+    return Status(ONNXRUNTIME, FAIL, "The node is not placed on any Execution Provider, therefore, can't find a suitable kernel for it");
   }
-
   Status status;
-  for (auto& registry : kernel_registries_) {
-    status = registry->CreateKernel(node,
-                                    execution_provider,
-                                    session_state.GetInitializedTensors(),
-                                    session_state.GetMLValueNameIdxMap(),
-                                    session_state.GetFuncMgr(),
-                                    op_kernel);
+  {
+    std::lock_guard<OrtMutex> lock(lock_);
+    for (auto& registry : kernel_registries_) {
+      status = registry->CreateKernel(node,
```

Comment:
>CreateKernel [](start = 25, length = 12)

Side question: would TryCreateKernel be a better name?

Reply:
Why?

Reply:
Because it only attempts to create the kernel if found by TryFindKernel. If I see CreateKernel as the name, I would not think failure was expected. But in this case we're attempting to create a kernel if found, and expect failures. 

---
In reply to: [256098369](https://github.com/Microsoft/onnxruntime/pull/465#discussion_r256098369) [](ancestors = 256098369)

Reply:
As you said, usually it shouldn't fail.

Reply:
I didn't intend to. :) What I meant was we only attempt to create it if TryFindKernel is successful. Calling this TryCreateKernel would be more consistent with that. If found, we don't expect it to fail, but lookups in custom registries would (I assume) fail more often than not. 

---
In reply to: [258171590](https://github.com/Microsoft/onnxruntime/pull/465#discussion_r258171590) [](ancestors = 258171590)

Reply:
This function is called after graph partition, after the kernel search. So, we expect we can find it.

Reply:
KernelRegistryManager::CreateKernel calls KernelRegistry::CreateKernel,which calls KernelRegistry::TryFindKernel.

KernelRegistry::TryFindKernel is also called in the graph partition phrase.